packages/stock-location/src/services/stock-location.ts

```diff
@@ -0,0 +1,266 @@
+   * @param {string} stockLocationId - The ID of the stock location.
+   * @param {FindConfig} [config={}] - Additional configuration for the query.
+   * @return {Promise<StockLocation>} The stock location.
+   * @throws {MedusaError} If the stock location ID is not defined.
+   * @throws {MedusaError} If the stock location with the given ID was not found.
+   */
+  async retrieve(
+    stockLocationId: string,
+    config: FindConfig<StockLocation> = {}
+  ): Promise<StockLocation> {
+    if (!isDefined(stockLocationId)) {
+      throw new MedusaError(
+        MedusaError.Types.NOT_FOUND,
+        `"stockLocationId" must be defined`
+      )
+    }
+
+    const manager = this.getManager()
+    const locationRepo = manager.getRepository(StockLocation)
+
+    const query = buildQuery({ id: stockLocationId }, config)
+    const loc = await locationRepo.findOne(query)
+
+    if (!loc) {
+      throw new MedusaError(
+        MedusaError.Types.NOT_FOUND,
+        `StockLocation with id ${stockLocationId} was not found`
+      )
+    }
+
+    return loc
+  }
+
+  /**
+   * Creates a new stock location.
+   * @param {CreateStockLocationInput} data - The input data for creating a stock location.
+   * @returns {Promise<StockLocation>} - The created stock location.
+   */
+  async create(data: CreateStockLocationInput): Promise<StockLocation> {
+    const defaultManager = this.getManager()
+    return await defaultManager.transaction(async (manager) => {
+      const locationRepo = manager.getRepository(StockLocation)
+
+      const loc = locationRepo.create({
+        name: data.name,
+      })
+
+      if (isDefined(data.address) || isDefined(data.address_id)) {
+        if (typeof data.address === "string" || data.address_id) {
+          const addrId = (data.address ?? data.address_id) as string
+          const address = await this.retrieve(addrId, {
+            select: ["id"],
+          })
+          loc.address_id = address.id
+        } else {
+          const locAddressRepo = manager.getRepository(StockLocationAddress)
+          const locAddress = locAddressRepo.create(data.address!)
+          const addressResult = await locAddressRepo.save(locAddress)
+          loc.address_id = addressResult.id
+        }
+      }
+
+      const { metadata } = data
+      if (metadata) {
+        loc.metadata = setMetadata(loc, metadata)
+      }
+
+      const result = await locationRepo.save(loc)
+
+      await this.eventBusService_.emit(StockLocationService.Events.CREATED, {
+        id: result.id,
+      })
+
+      return result
+    })
+  }
+
+  /**
+   * Updates an existing stock location.
+   * @param {string} itemId - The ID of the stock location to update.
+   * @param {UpdateStockLocationInput} updateData - The update data for the stock location.
+   * @returns {Promise<StockLocation>} - The updated stock location.
+   */
+
+  async update(
+    itemId: string,
+    updateData: UpdateStockLocationInput
```

Comment:
Which consumer you mean? As this will be an independent module, a transaction orchestrator will be needed to revert failed transaction involving this module and the core.

Reply:
only on the JS client, there is a * @experimental on the JSDoc